Source/Core/Heap.js

```diff
@@ -0,0 +1,200 @@
+            },
+            set : function(value) {
+                this._maximumLength = value;
+                if (this._length > value && value > 0) {
+                    this._length = value;
+                    this._array.length = value;
+                }
+            }
+        }
+    });
+
+    function swap(array, a, b) {
+        var temp = array[a];
+        array[a] = array[b];
+        array[b] = temp;
+    }
+
+    /**
+     * Resizes the internal array of the heap.
+     *
+     * @param {Number} [length] The length to resize internal array to. Defaults to the current length of the heap.
+     */
+    Heap.prototype.reserve = function(length) {
+        length = defaultValue(length, this._length);
+        this._array.length = length;
+    };
+
+    /**
+     * Update the heap so that index and all descendants satisfy the heap property.
+     *
+     * @param {Number} [index=0] The starting index to heapify from.
+     */
+    Heap.prototype.heapify = function(index) {
+        index = defaultValue(index, 0);
+        var length = this._length;
+        var comparator = this._comparator;
+        var array = this._array;
+        var candidate = -1;
+
+        while (true) {
+            var right = 2 * (index + 1);
+            var left = right - 1;
+
+            if (left < length && comparator(array[left], array[index]) < 0) {
+                candidate = left;
+            } else {
+                candidate = index;
+            }
+
+            if (right < length && comparator(array[right], array[candidate]) < 0) {
+                candidate = right;
+            }
+            if (candidate !== index) {
+                swap(array, candidate, index);
+                index = candidate;
+            } else {
+                break;
+            }
+        }
+    };
+
+    /**
+     * Insert an element into the heap. If the length would grow greater than maximumLength
+     * of the heap, extra elements are removed.
+     *
+     * @param {*} element The element to insert
+     *
+     * @return {*} The element that was removed from the heap if the heap is at full capacity.
+     */
+    Heap.prototype.insert = function(element) {
+        //>>includeStart('debug', pragmas.debug);
+        Check.defined('element', element);
+        //>>includeEnd('debug');
+
+        var array = this._array;
+        var comparator = this._comparator;
+
+        var index = this._length++;
+        if (index < array.length) {
+            array[index] = element;
+        } else {
+            array.push(element);
+        }
+
+        while (index !== 0) {
+            var parent = Math.floor((index - 1) / 2);
+            if (comparator(array[index], array[parent]) < 0) {
+                swap(array, index, parent);
+                index = parent;
+            } else {
+                break;
+            }
+        }
+
+        var removedElement;
+
+        if (defined(this._maximumLength) && (this._length > this._maximumLength)) {
```

Comment:
both `this._length` and `this._maximumLength` are accessed multiple times throughout this function, store them in locals.